lib/internal/streams/whatwg.js

```diff
@@ -0,0 +1,4233 @@
+  controller,
+  desc) {
+  const {
+    buffer,
+    byteLength,
+    byteOffset,
+    bytesFilled,
+    elementSize,
+  } = desc;
+  const currentAlignedBytes = bytesFilled - (bytesFilled % elementSize);
+  const maxBytesToCopy = MathMin(
+    controller[kState].queueTotalSize,
+    byteLength - bytesFilled);
+  const maxBytesFilled = bytesFilled + maxBytesToCopy;
+  const maxAlignedBytes = maxBytesFilled - (maxBytesFilled % elementSize);
+  let totalBytesToCopyRemaining = maxBytesToCopy;
+  let ready = false;
+  if (maxAlignedBytes > currentAlignedBytes) {
+    totalBytesToCopyRemaining = maxAlignedBytes - bytesFilled;
+    ready = true;
+  }
+  const {
+    queue,
+  } = controller[kState];
+
+  while (totalBytesToCopyRemaining) {
+    const headOfQueue = queue[0];
+    const bytesToCopy = MathMin(
+      totalBytesToCopyRemaining,
+      headOfQueue.byteLength);
+    const destStart = byteOffset + desc.bytesFilled;
+    copyArrayBuffer(
+      buffer,
+      destStart,
+      headOfQueue.buffer,
+      headOfQueue.byteOffset,
+      bytesToCopy);
+    if (headOfQueue.byteLength === bytesToCopy) {
+      ArrayPrototypeShift(queue);
+    } else {
+      headOfQueue.byteOffset += bytesToCopy;
+      headOfQueue.byteLength -= bytesToCopy;
+    }
+    controller[kState].queueTotalSize -= bytesToCopy;
+    readableByteStreamControllerFillHeadPullIntoDescriptor(
+      controller,
+      bytesToCopy,
+      desc);
+    totalBytesToCopyRemaining -= bytesToCopy;
+  }
+
+  if (!ready) {
+    assert(controller[kState].queueTotalSize === 0);
+    assert(desc.bytesFilled > 0);
+    assert(desc.bytesFilled < elementSize);
+  }
+  return ready;
+}
+
+function readableByteStreamControllerProcessPullIntoDescriptorsUsingQueue(
+  controller) {
+  const {
+    closeRequested,
+    pendingPullIntos,
+    stream,
+  } = controller[kState];
+  assert(!closeRequested);
+  while (pendingPullIntos.length) {
+    if (controller[kState].queueTotalSize === 0)
+      return;
+    const desc = pendingPullIntos[0];
+    if (readableByteStreamControllerFillPullIntoDescriptorFromQueue(
+      controller,
+      desc)) {
+      readableByteStreamControllerShiftPendingPullInto(controller);
+      readableByteStreamControllerCommitPullIntoDescriptor(stream, desc);
+    }
+  }
+}
+
+function readableByteStreamControllerRespondInReadableState(
+  controller,
+  bytesWritten,
+  desc) {
+  const {
+    buffer,
+    bytesFilled,
+    byteLength,
+  } = desc;
+  assert(bytesFilled + bytesWritten <= byteLength);
+  readableByteStreamControllerFillHeadPullIntoDescriptor(
+    controller,
+    bytesWritten,
+    desc);
+
+  if (desc.bytesFilled < desc.elementSize)
+    return;
+
+  readableByteStreamControllerShiftPendingPullInto(controller);
+
+  const remainderSize = desc.bytesFilled % desc.elementSize;
+
+  if (remainderSize) {
+    const end = desc.byteOffset + desc.bytesFilled;
+    const remainder = buffer.slice(end - remainderSize, remainderSize);
```

Comment:
do we cache Buffer slice?

Reply:
As far as I know, no.

Reply:
Might want to make a new buffer and copy the slice in order to avoid a long living reference to pool.

Reply:
@ronag:
> Might want to make a new buffer and copy the slice in order to avoid a long living reference to pool.

There's no need. The buffer at this point is not connected to the pool.